File: sdk/php/tests/WorkerTest.php
<?php

declare(strict_types=1);

namespace Cordum\Cap\Tests;

use Cordum\Agent\V1\JobRequest;
use Cordum\Agent\V1\JobResult;
use Cordum\Agent\V1\JobStatus;
use Cordum\Cap\Context;
use Cordum\Cap\MiddlewareChain;
use Cordum\Cap\TestHelper;
use PHPUnit\Framework\TestCase;

class WorkerTest extends TestCase
{
    #[\PHPUnit\Framework\Attributes\Test]
    public function submitAndWaitReturnsResult(): void
    {
        $handler = function (Context $ctx, JobRequest $req): JobResult {
            $result = new JobResult();
            $result->setJobId($req->getJobId());
            $result->setStatus(JobStatus::JOB_STATUS_SUCCEEDED);
            $result->setWorkerId('test-worker');
            return $result;
        };

        $req = new JobRequest();
        $req->setJobId('j-1');
        $req->setTopic('test');

        $result = TestHelper::submitAndWait($handler, $req);
        $this->assertSame('j-1', $result->getJobId());
        $this->assertSame(JobStatus::JOB_STATUS_SUCCEEDED, $result->getStatus());
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function submitAndWaitContextPopulated(): void
    {
        $captured = null;
        $handler = function (Context $ctx, JobRequest $req) use (&$captured): JobResult {
            $captured = $ctx;
            $result = new JobResult();
            $result->setJobId($req->getJobId());
            $result->setStatus(JobStatus::JOB_STATUS_SUCCEEDED);
            $result->setWorkerId('w');
            return $result;
        };

        $req = new JobRequest();
        $req->setJobId('j');
        $req->setTopic('t');

        TestHelper::submitAndWait($handler, $req);

        $this->assertSame('test-trace', $captured->traceId);
        $this->assertSame('test-sender', $captured->senderId);
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function middlewareFIFOOrder(): void
    {
        $order = [];

        $makeMw = function (int $id) use (&$order): callable {
            return function (callable $next) use ($id, &$order): callable {
                return function (Context $ctx, JobRequest $req) use ($next, $id, &$order): JobResult {
                    $order[] = $id;
                    return $next($ctx, $req);
                };
            };
        };

        $baseHandler = function (Context $ctx, JobRequest $req): JobResult {
            $result = new JobResult();
            $result->setJobId($req->getJobId());
            $result->setStatus(JobStatus::JOB_STATUS_SUCCEEDED);
            $result->setWorkerId('w');
            return $result;
        };

        $wrapped = MiddlewareChain::apply($baseHandler, [$makeMw(1), $makeMw(2), $makeMw(3)]);

        $ctx = new Context(traceId: 't', senderId: 's');
        $req = new JobRequest();
        $req->setJobId('j');
        $req->setTopic('t');
        $wrapped($ctx, $req);

        $this->assertSame([1, 2, 3], $order);
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function emptyMiddlewareReturnsBase(): void
    {
        $baseHandler = function (Context $ctx, JobRequest $req): JobResult {
            $result = new JobResult();
            $result->setJobId($req->getJobId());
            $result->setStatus(JobStatus::JOB_STATUS_SUCCEEDED);
            $result->setWorkerId('w');
            return $result;
        };

        $wrapped = MiddlewareChain::apply($baseHandler, []);
        $ctx = new Context(traceId: 't', senderId: 's');
        $req = new JobRequest();
        $req->setJobId('j-1');
        $req->setTopic('t');

        $result = $wrapped($ctx, $req);
        $this->assertSame('j-1', $result->getJobId());
    }

    #[\PHPUnit\Framework\Attributes\Test]
    public function loggingMiddlewareDoesNotInterfere(): void
    {
        $baseHandler = function (Context $ctx, JobRequest $req): JobResult {
            $result = new JobResult();
            $result->setJobId($req->getJobId());
            $result->setStatus(JobStatus::JOB_STATUS_SUCCEEDED);
            $result->setWorkerId('w');
            return $result;
        };

        $wrapped = MiddlewareChain::apply($baseHandler, [MiddlewareChain::logging()]);
        $ctx = new Context(traceId: 't', senderId: 's');
        $req = new JobRequest();
        $req->setJobId('logged-job');
        $req->setTopic('t');

        $result = $wrapped($ctx, $req);
        $this->assertSame('logged-job', $result->getJobId());
    }
}

File: sdk/java/src/test/java/io/cordum/cap/WorkerTest.java
package io.cordum.cap;

import io.cordum.cap.agent.v1.JobRequest;
import io.cordum.cap.agent.v1.JobResult;
import io.cordum.cap.agent.v1.JobStatus;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class WorkerTest {

    @Test
    void handlerInvocation() throws Exception {
        JobRequest req = JobRequest.newBuilder()
                .setJobId("job-1").setTopic("test").build();

        JobResult result = Testing.submitAndWait((ctx, r) ->
                JobResult.newBuilder()
                        .setJobId(r.getJobId())
                        .setStatus(JobStatus.JOB_STATUS_SUCCEEDED)
                        .build(), req);

        assertEquals("job-1", result.getJobId());
        assertEquals(JobStatus.JOB_STATUS_SUCCEEDED, result.getStatus());
    }

    @Test
    void middlewareOrdering() throws Exception {
        List<Integer> order = new ArrayList<>();

        Middleware.Interceptor mw1 = next -> (ctx, req) -> {
            order.add(1);
            return next.handle(ctx, req);
        };
        Middleware.Interceptor mw2 = next -> (ctx, req) -> {
            order.add(2);
            return next.handle(ctx, req);
        };

        Middleware.Handler base = (ctx, req) -> {
            order.add(3);
            return JobResult.newBuilder()
                    .setJobId(req.getJobId())
                    .setStatus(JobStatus.JOB_STATUS_SUCCEEDED)
                    .build();
        };

        Middleware.Handler wrapped = Middleware.applyMiddleware(base, List.of(mw1, mw2));

        JobRequest req = JobRequest.newBuilder().setJobId("j").setTopic("t").build();
        wrapped.handle(new Middleware.Context("t", "s", null), req);

        assertEquals(List.of(1, 2, 3), order);
    }

    @Test
    void emptyMiddlewareListReturnsBase() throws Exception {
        Middleware.Handler base = (ctx, req) ->
                JobResult.newBuilder()
                        .setJobId(req.getJobId())
                        .setStatus(JobStatus.JOB_STATUS_SUCCEEDED)
                        .build();

        Middleware.Handler wrapped = Middleware.applyMiddleware(base, List.of());

        JobRequest req = JobRequest.newBuilder().setJobId("j-1").setTopic("t").build();
        JobResult result = wrapped.handle(new Middleware.Context("t", "s", null), req);
        assertEquals("j-1", result.getJobId());
    }

    @Test
    void contextFieldsPopulated() throws Exception {
        final String[] captured = new String[2];

        Middleware.Handler handler = (ctx, req) -> {
            captured[0] = ctx.traceId();
            captured[1] = ctx.senderId();
            return JobResult.newBuilder()
                    .setJobId(req.getJobId())
                    .setStatus(JobStatus.JOB_STATUS_SUCCEEDED)
                    .build();
        };

        JobRequest req = JobRequest.newBuilder().setJobId("j").setTopic("t").build();
        handler.handle(new Middleware.Context("my-trace", "my-sender", null), req);

        assertEquals("my-trace", captured[0]);
        assertEquals("my-sender", captured[1]);
    }
}
